### ISSUE TITLE
SIGILL while returning from usermode

### ISSUE THREAD

**Issue by avinashpqtn** [foss-for-synopsys-dwc-arc-processors/toolchain#91]
Hi,

We are porting Linux kernel on our internal platform to 4.7(gcc version 4.8.5 (ARCompact ISA Linux uClibc toolchain built on 20160906) ).

I 

We see this peculiar crash where init is launched from kernel boot but killed immediately. I tried with my sample init which prints messages to console and returns. SIGILL is seen in this case as well.
I think this SIGILL is caused because of stack corruption and is seen after returning from usermode.

Console log also has information about TASK_SIZE, VMALLOC regions etc.

Do you see any issue with attached config file which maybe causing this crash? 

[arc_4.7_config.txt](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/528196/arc_4.7_config.txt)
[stacktrace_log2.txt](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/528197/stacktrace_log2.txt)

**Comment by vineetgarc** [comment]
why do you need to tinker with TASK_SIZE and VMALLOC_SIZE. The vmalloc size you have would anyways fit in the default settings @ 0x7000_0000.
Can you also check if this start address is aligned to PMD_SIZE

**Comment by vineetgarc** [comment]
and BTW, CONFIG_ARC_KVADDR_SIZE=48 means 48MB of kvaddr space. This would be too less for a typical kernel run, considering loadable modules also use the same vaddr space.

**Comment by vineetgarc** [comment]
While I don't understand you made the changes, I did a quick build of a kernel with CONFIG_ARC_KVADDR_SIZE=48 and the hacks below (which u must be having as well) and it seems to boot fine for me !

-#define TASK_SIZE      0x60000000
+#define TASK_SIZE      0x54000000

-#define VMALLOC_START  (PAGE_OFFSET - (CONFIG_ARC_KVADDR_SIZE << 20))
+#define VMALLOC_START  0x55000000

**Comment by avinashpqtn** [comment]
Hi Vineet,

Thanks for information about boot being successful in your setup.
I made changes in config file to increase KVADDR_SIZE to 256(default); I still see the crash. 

So this maynot be related to kernel memory. Looking at stacktrace logs, this looks like PC is moving to some illegal address and starting to execute from there and sees that as illegal instruction. Do you think that return to kernel space is causing stack corruption?

**Comment by vineetgarc** [comment]
hard to say really, BTW u didn't answer my question about need to change these things in first place. You can send me pvt email if this is confidential info. Do things work fine with default values of these #defines ?

**Comment by avinashpqtn** [comment]
Please give me your email ID. We can talk over emails.

**Comment by vineetgarc** [comment]
vgupta@synopsys.com
BTW I', based in Mt. View so you can also call me @ 650 584 4692

**Comment by avinashpqtn** [comment]
Sure! We can talk over as well...

**Comment by vineetgarc** [comment]
The userspace crash was due to instruction error exception for LLOCK/SCOND which ARC750 lacks. This was coming from uClibc NPTL__arch_compare_and_exchange_val_32_acq() which only has  llock/scond based implementation.There are 2 possible solutions: (1) using EX based atomic swap to build this primitive, which uses an additional central spinlock to provide atomicity - but this can cause  tight spin loops in case of contention (2) use a new syscall to provide kernel assist - which is safer - but causes entry into kernel for each such operation hence potentially costly. I'm experimenting with both approaches at the moment.

**Comment by vineetgarc** [comment]
Hi Avinash,
Since you agreed to be the guinea pig ;-) I'm attaching a prelim patch to unblock you. It seems to work fine in my limited tests so far. In near future, this will be replaced with a better/complex fix which I'm still testing/ working on.
[0001-NPTL-ARC-donot-merge-provide-an-EX-based-atomix-cmpx.patch.txt](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/542596/0001-NPTL-ARC-donot-merge-provide-an-EX-based-atomix-cmpx.patch.txt)

To apply, in your uclibc tree, 
patch -p1 < ./0001-NPTL-ARC-donot-merge-provide-an-EX-based-atomix-cmpx.patch.txt 

and then rebuild the whole toolchain

**Comment by avinashpqtn** [comment]
Thanks, Vineet.
Patch applied and now I have started recompilation. Will update about results soon.

**Comment by avinashpqtn** [comment]
Hi Vineet,
Early update: Now I dont see LLOCK instructions in objdump of uClibc and we dont encounter earlier mentioned crash. But- we are hitting a new illegal instruction crash now this one points to "swape".
I have disabled -mswape option in Rules.mak and I think we are still seeing mswape code.

Here is portion of CPU flags related to this:
ifeq ($(TARGET_ARCH),arc)

# #      CPU_CFLAGS-y += -mswape

```
    CPU_CFLAGS-$(CONFIG_ARC_CPU_700) += -mA7
    CPU_CFLAGS-$(CONFIG_ARC_CPU_HS) += -mcpu=archs -mlock
    CPU_LDFLAGS-y += $(CPU_CFLAGS) -marclinux
```

endif

I am reinstalling toolchain just to be sure and will update you once more in sometime.

**Comment by vineetgarc** [comment]
Can u try hacking libc/sysdeps/linux/arc/bits/byteswap.h - so that the swape instruction is not compiled it - i'm looking into whats going on there

**Comment by vineetgarc** [comment]
i removed -mswape and I don't see the instruction there anymore - r u sure u did the rebuild and/or copied the new library bins to target !

**Comment by avinashpqtn** [comment]
1. For now, I have hacked byteswap.h for not adding swape instruction.
2. After this compilation, I will clean everything including destination directories, remove this hack and just disable -mswape from Rules.mak and will see if swape instructions are added.

**Comment by vineetgarc** [comment]
Hi Avinash, Can you confirm if your userland comes up fine after those hacks ?

**Comment by avinashpqtn** [comment]
Hi Vineet,

I confirm that uClibc patch provided by you enables usermode to start successfully.

**Comment by vineetgarc** [comment]
Hi Avinash,

Considering this works for you - I'm closing this issue. Note that the upcoming GNU 2016.09 release of tools has a proper uClibc fix for this, so you may wanna consider moving to it for your development going fwd.

Do note that this relies on the syscall assist added to v4.9 kernel which unfortunately can't be backported so you might have to cherry-pick upstream fixes below to your kernel (v4.7?)

2016-11-07 e6e335bf3a40 ARC: change return value of userspace cmpxchg assist syscall  
2016-10-20 91e040a79df7 ARC: syscall for userspace cmpxchg assist  
2016-08-10 840c054fd0ef ARC: Support syscall ABI v4
